Adds test framework and an example test scripts.

sawsa307 opened the pull request:
* Adds test framework and an example of recipe test scripts
* Adds instructions to run the recipe tests in the README

sawsa307 commented on test/recipe_test.go:
```diff
@@ -0,0 +1,69 @@
+				t.Parallel()
+				setupFile := path.Join(fp, dir.Name(), "setup.sh")
+				if _, err := os.Stat(setupFile); err != nil {
+					return
```
We talked about that we will skip running test for directories that don't have scripts(setup.sh/cleanup.sh/run-test.sh) so that's why I added these.
Do we still want to treat them as ignored or raise an error?

bowei replied:
We should give some kind of indication that the test is being skipped because of the scripts missing.

I think you can use t.Skip() to signal this.

sawsa307 replied:
Updated and tested.

sawsa307 commented on test/cleanup_all.sh:
```diff
@@ -0,0 +1,20 @@
+
+dir="ingress/single-cluster/*/cleanup.sh"
+for cleanup_file in $dir; do
+    bash "${cleanup_file}"
```
Added.

sawsa307 commented on test/cleanup_all.sh:
```diff
@@ -0,0 +1,20 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+dir="ingress/single-cluster/*/cleanup.sh"
```
Yes that makes sense. Added.

sawsa307 commented on test/README.md:
````diff
@@ -26,20 +26,32 @@ Make sure you have valid credentials for accessing GCP:
 
 ```
-export LOCATION=<your-location>
+bash ingress/single-cluster/ingress-external-basic/setup.sh
````
Updated.

sawsa307 commented on test/README.md:
````diff
@@ -26,20 +26,32 @@ Make sure you have valid credentials for accessing GCP:
+
+To cleanup a specific test separately, you can run its cleanup.sh.
+```
+bash ingress/single-cluster/ingress-external-basic/cleanup.sh
````
Updated.

bowei commented on test/recipe_test.go:
```diff
@@ -0,0 +1,69 @@
+			dir := dir
+			t.Run(dir.Name(), func(t *testing.T) {
+				t.Parallel()
+				setupFile := path.Join(fp, dir.Name(), "setup.sh")
```
You can use a for loop to avoid redundant code.

**do not** assume err means not found/missing. You should print the actual error that is being return from stat()

```
for _, file := range []string{"setup.sh", "run-test.sh", "cleanup.sh"} {
  path := path.Join(fp, dir.Name(), file)
  if _, err := os.Stat(file); err != nil {
    t.Skipf("Skipping test %q: stat(%q) = %v", dir.Name(), path, err)
  }
}

```

bowei commented on test/recipe_test.go:
```diff
@@ -0,0 +1,69 @@
+			t.Fatalf("Failed to open directory %s: %v", fp, err)
+		}
+
+		for _, dir := range dirs {
```
this should be factored into a separate func

You should stat() `dir` to make sure it is actually a directory.

bowei replied:
dir => fileName

bowei commented on test/recipe_test.go:
```diff
@@ -0,0 +1,69 @@
+
+func TestRecipe(t *testing.T) {
+	for _, fp := range testFilePaths {
+		dirs, err := os.ReadDir(fp)
```
dirs should be called "fileNames"

bowei commented on test/recipe_test.go:
```diff
@@ -0,0 +1,80 @@
+	for _, file := range []string{"setup.sh", "run-test.sh", "cleanup.sh"} {
+		path := path.Join(recipeDir, file)
+		if _, err := os.Stat(path); err != nil {
+			t.Skipf("Skipping test %q: stat(%q) = %v", recipeDir, path, err)
```
I'm not sure this is quite what we want to do.

We shouldn't partially the the scripts if not all of them exist.

bowei replied:
You should check the existence of ALL before you proceed to run the scripts.

sawsa307 replied:
Updated.

sawsa307 commented on test/recipe_test.go:
```diff
@@ -0,0 +1,84 @@
+	for _, path := range paths {
+		out, err := exec.Command("bash", path).CombinedOutput()
+		if err != nil {
+			t.Fatalf("Test %s failed when running %q: %q, err: %v", recipeDir, path, out, err)
```
Fail now because we shouldn't continue testing if any step fails.

bowei commented on test/recipe_test.go:
```diff
@@ -0,0 +1,85 @@
+	for _, fp := range testFilePaths {
+		fileNames, err := os.ReadDir(fp)
+		if err != nil {
+			t.Fatalf("Failed to open directory %s: %v", fp, err)
```
Don't overcomplicate error messages

t.Fatalf("os.ReadDir(%q) = %v", fp, err)

bowei reviewed (commented):
Fix the commit comment 

```
Adds test framework and an example test scripts

* Adds test framework and an example test scripts
* Adds instructions to the run the tests in the README
```

bowei commented:
Can you make the following changes to the Makefile
Also, fix the copyright in the Makefile

```
diff --git a/Makefile b/Makefile
index 2a89aef..10d9f68 100644
--- a/Makefile
+++ b/Makefile
@@ -15,11 +15,14 @@
 RUN_IN_PROW ?= false
 NUM_NODES ?= 3
 TEST_TO_RUN ?= .*
+TEST_GOFILES := $(shell find ./test -name \*.go)
 
 all: bin/recipes-test
 
-bin/recipes-test:
-       mkdir bin/
+bin:
+       mkdir ./bin
+
+bin/recipes-test: bin $(TEST_GOFILES)
        go test -c -o $@ ./test
 
 .PHONY: test
@@ -35,4 +38,4 @@ cleanenv:
 
 .PHONY: clean
 clean:
-       rm -rf bin/
+       rm -rf ./bin
```

sawsa307 commented:
> Can you make the following changes to the Makefile Also, fix the copyright in the Makefile
> 
> ```
> diff --git a/Makefile b/Makefile
> index 2a89aef..10d9f68 100644
> --- a/Makefile
> +++ b/Makefile
> @@ -15,11 +15,14 @@
>  RUN_IN_PROW ?= false
>  NUM_NODES ?= 3
>  TEST_TO_RUN ?= .*
> +TEST_GOFILES := $(shell find ./test -name \*.go)
>  
>  all: bin/recipes-test
>  
> -bin/recipes-test:
> -       mkdir bin/
> +bin:
> +       mkdir ./bin
> +
> +bin/recipes-test: bin $(TEST_GOFILES)
>         go test -c -o $@ ./test
>  
>  .PHONY: test
> @@ -35,4 +38,4 @@ cleanenv:
>  
>  .PHONY: clean
>  clean:
> -       rm -rf bin/
> +       rm -rf ./bin
> ```

Updated. Also deleted TEST_TO_RUN since we only have one test now.

bowei reviewed (approved):
lgtm
